service/storage_service.cc

```diff
@@ -2037,8 +2037,40 @@ class topology_coordinator {
         slogger.info("raft topology: coordinator fiber found a node to work on id={} state={}", node.id, node.rs->state);
 
         switch (node.rs->state) {
-            case node_state::normal:
             case node_state::none: {
+                if (_topo_sm._topology.normal_nodes.empty()) {
+                    slogger.info("raft topology: skipping join node handshake for the first node in the cluster");
+                } else {
+                    auto validation_result = validate_joining_node(node);
+                    if (auto* reject = std::get_if<join_node_response_params::rejected>(&validation_result)) {
+                        // Find the IP first.
+                        auto ip = id2ip(locator::host_id(node.id.uuid()));
```

Comment:
BTW. `id2ip` is a temporary helper, it retries forever. Once we migrate to host IDs for identification, any translation attempts we do should only happen for the purposes of messaging (like here) and should be protected with timeouts. I would thus refrain from using `id2ip`, keep using it only in these temporary use cases and not in cases like this one.

Reply:
> Why do it first? Should failure to do the translation result in failure to reject node boot?

In tests, I observed that this IP is usually unavailable at the beginning and it usually takes 2 attempts for the gossiped IP to reach the topology coordinator. I do the translation before transition to increase the chance of success of the `join_node_response` rpc that follows `update_topology_state`. Though, in all fairness, I should probably use `_gossiper.wait_alive` for that purpose.

> BTW. id2ip is a temporary helper, it retries forever. Once we migrate to host IDs for identification, any translation attempts we do should only happen for the purposes of messaging (like here) and should be protected with timeouts.

Ok, in that case it's another argument in favor of `_gossiper.wait_alive` before `update_topology_state`. Alternatively, I can add another state that marks a node for rejection. See my comment here: https://github.com/scylladb/scylladb/pull/15196#discussion_r1329841402

Reply:
> BTW. `id2ip` is a temporary helper, it retries forever. Once we migrate to host IDs for identification, any translation attempts we do should only happen for the purposes of messaging (like here) and should be protected with timeouts. I would thus refrain from using `id2ip`, keep using it only in these temporary use cases and not in cases like this one.

Messaging should be done using ids as well, so the use of `id2ip` here is temporary as well.

Reply:
> In fact, failure translation could easily happen if the bootstrapping node dies and some time passes before we reach this step.

But we decided in such a case the bootstrap will not restart, so it is fine to fail it.

Reply:
> In tests, I observed that this IP is usually unavailable at the beginning and it usually takes 2 attempts for the gossiped IP to reach the topology coordinator. 

We should not rely on retry to wait for gossiper to converge any more. This will just complicate error handling (should you fail or retry?).

Reply:
Now, `id2ip` is put just before the RPC call and there is a retry loop with a timeout that tries to send the response.